library/init/data/nat/lemmas.lean

```diff
@@ -730,5 +770,148 @@ lemma pred_inj : ∀ {a b : nat}, a > 0 → b > 0 → nat.pred a = nat.pred b 
 
 /- TODO(Leo): sub + inequalities -/
 
+/- div -/
+
+theorem zero_div : ∀ m : ℕ, 0 / m = 0
+  | zero := by refl
+  | (succ n) := by refl
+
+theorem div_le_div
+: ∀ {y x}, x ≤ y → ∀ z : ℕ, x / z ≤ y / z :=
+  take y,
+  let P y := ∀ {x}, x ≤ y → ∀ z : ℕ, x / z ≤ y / z in
+  @well_founded.induction _ _ (measure_wf id) P y $
+    take y,
+    assume IH : ∀ (y' : ℕ), y' < y → P y',
+    take x,
+    assume h' : x ≤ y,
+    take z,
+    if h : 0 < z ∧ z ≤ x then
+      have Hz : 0 < z, from h^.left,
+      have Hx : 0 < z ∧ z ≤ x,    from h,
+      have Hy : 0 < z ∧ z ≤ y,    from ⟨Hz,nat.le_trans h^.right h'⟩,
+      have Hxy_z : x - z ≤ y - z, from nat.sub_le_sub_right h' z,
+      have Hyz_z : y - z < y,     from nat.sub_lt (lt_of_lt_of_le Hz Hy^.right) Hz,
+      begin
+        rw [div_def x z,dif_pos Hx,div_def y z,dif_pos Hy,add_le_add_right_iff] ,
+        apply IH (y - z) Hyz_z Hxy_z
+      end
+    else
+      have h' : ¬ 0 < z ∨ ¬ z ≤ x,
+         from classical.not_and_of_not_or_not h,
+      begin
+        cases h' with h₀ h₀,
+        { note h₁ := not_pos_of_eq_zero h₀,
+          simp [h₁], cases x, cases y, refl, refl, cases y, refl, refl },
+        { assert Hx : ¬ (0 < z ∧ z ≤ x),
+          { intro HH, apply (h₀ HH^.right) },
+            rw [div_def x z,dif_neg Hx], apply zero_le }
+     end
+
+theorem div_pred_to_pred_div
+: ∀ (x : ℕ) {z : ℕ}, 1 ≤ z → (z * x - 1) / z = x - 1
+ | zero z := assume _, begin cases z, refl, change 0 / succ a = 0 - 1, refl end
+ | (succ x) z :=
+    assume Pz : 1 ≤ z,
+    if Hpos : 1 ≤ x then
+    have Pxz : 1 ≤ z * x,
+      from nat.mul_le_mul Pz Hpos,
+    have Pz_le_zx_z_1 : z ≤ z * x + z - 1, from calc
+         z * x + z - 1
+       = z * x + (z - 1) : by rw -nat.add_sub_assoc Pz
+...    ≥ 1 + (z - 1)     : nat.add_le_add_right Pxz _
+...    = 1 + z - 1       : by rw -nat.add_sub_assoc Pz
+...    = z               : by rw nat.add_sub_cancel_left,
+    have Pdiv : 0 < z ∧ z ≤ z * x + z - 1, from ⟨Pz,Pz_le_zx_z_1⟩,
+    calc
```

Comment:
I'd recommend a final pass to see if these proofs can be simplified via the tactics language.